scripts/generate-components.js

```diff
@@ -104,6 +110,7 @@ const ${Component} = (props) => {
                 onClick={() => {
                     if (props.setProps) props.setProps({n_clicks: props.n_clicks + 1});
                     if (props.fireEvent) props.fireEvent({event: 'click'});
+                    if (props.setProps) props.setProps({n_clicks_previous: props.n_clicks_previous + 1});
```

Comment:
I actually don't quite understand this - how is `n_clicks_previous` just not equal to `n_clicks`?

Reply:
`setProps` is the `dash-renderer` provided function which will propagate these changes down to Dash's backend and then causes the component / app to re-render. I'm a little wary of calling `setProps` _twice_ on an `onClick` handler as I would expect that this would fire the callbacks twice and cause multiple re-renders. In practice, the callback might not be called twice because `dash-renderer` does some things to trim an excessive number of requests (https://github.com/plotly/dash-renderer/pull/22).

Instead, what if we did something like:
```
const newProps = {
    n_clicks: props.n_clicks + 1
}
if (newProps.n_clicks > 1) {
    newProps.n_clicks_previous = props.n_clicks_previous + 1;
}
setProps(newProps)
```

Reply:
I tried something similar, but I found that it has equivalent behavior. When there are two properties in newProps and both of those properties are inputs of a callback, the callback is triggered twice.

For example, setting the innards of `generate-component.js` to 
```
return (
            <${element}
                onClick={() => {
                    const newProps = {
                        n_clicks: props.n_clicks + 1
                    }
                    if (newProps.n_clicks > 1) {
                        newProps.n_clicks_previous = props.n_clicks_previous + 1
                    }
                    if (props.setProps) props.setProps(newProps);
                    if (props.fireEvent) props.fireEvent({event: 'click'});
                }}
                {...props}
            >
                {props.children}
            </${element}>
);
```
Generating the components, then testing with the test cases in `test_integration.py` as
```
def test_click(self):
        call_count = Value('i', 0)

        app = dash.Dash()
        app.layout = html.Div([
            html.Div(id='container'),
            html.Button('Click', id='button', n_clicks=0, n_clicks_previous=0)
        ])

        output_string = "You have clicked the button {} times." + \
                        "Previously, you have clicked the button {} times."

        @app.callback(Output('container', 'children'),
                      [Input('button', 'n_clicks'),
                       Input('button', 'n_clicks_previous')])
        def update_output(n_clicks, n_clicks_previous):
            call_count.value += 1
            print(n_clicks, n_clicks_previous)
            return output_string.format(n_clicks, n_clicks_previous)

        self.startServer(app)

        self.wait_for_element_by_css_selector('#container')

        self.wait_for_text_to_equal(
            '#container', output_string.format(0, 0))
        self.assertEqual(call_count.value, 1)
        self.snapshot('button initialization')

        self.driver.find_element_by_css_selector('#button').click()

        self.wait_for_text_to_equal(
            '#container', output_string.format(1, 0))
        self.assertEqual(call_count.value, 2)
        self.snapshot('button click one')

        self.driver.find_element_by_css_selector('#button').click()

        self.wait_for_text_to_equal(
            '#container', output_string.format(2, 1))
        self.assertEqual(call_count.value, 3)
        self.snapshot('button click two')

```

Yields the following testing error (while printing `n_clicks` and `n_clicks_previous` with each callback)

```
py27 runtests: commands[1] | python -m unittest tests.test_dash_html_components
..
----------------------------------------------------------------------
Ran 2 tests in 0.001s

OK
py27 runtests: commands[2] | python -m unittest tests.test_integration
 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
127.0.0.1 - - [03/Apr/2018 21:58:54] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:54] "GET /_dash-component-suites/dash_renderer/react@15.4.2.min.js?v=0.12.1 HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:54] "GET /_dash-component-suites/dash_renderer/react-dom@15.4.2.min.js?v=0.12.1 HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:54] "GET /_dash-component-suites/dash_renderer/bundle.js?v=0.12.1 HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:54] "GET /_dash-component-suites/dash_html_components/bundle.js?v=0.10.0 HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:55] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:55] "GET /_dash-dependencies HTTP/1.1" 200 -
(0, 0)  # Initialization of the component
127.0.0.1 - - [03/Apr/2018 21:58:55] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [03/Apr/2018 21:58:55] "GET /favicon.ico HTTP/1.1" 200 -
(1, 0) # First click -- only one property (props.n_clicks) is changed
127.0.0.1 - - [03/Apr/2018 21:58:56] "POST /_dash-update-component HTTP/1.1" 200 -
(2, 1) # Second click -- both props.n_clicks and props.n_clicks_previous are changed
127.0.0.1 - - [03/Apr/2018 21:58:56] "POST /_dash-update-component HTTP/1.1" 200 -
(2, 1) # Callback is triggered again
127.0.0.1 - - [03/Apr/2018 21:58:56] "POST /_dash-update-component HTTP/1.1" 200 -
F
======================================================================
FAIL: test_click (tests.test_integration.Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_integration.py", line 98, in test_click
    self.assertEqual(call_count.value, 3)
AssertionError: 4 != 3

----------------------------------------------------------------------
Ran 1 test in 15.905s

FAILED (failures=1)
ERROR: InvocationError for command '/home/ryan/MSK/dash/dash-html-components/.tox/py27/bin/python -m unittest tests.test_integration' (exited with code 1)
_______________________________________________ summary ________________________________________________
ERROR:   py27: commands failed

```
Thus, it only behaves correctly if `n_clicks_previous` is a State variable instead of Input, in which case it only matters what value `n_clicks_previous` has when setProps is called on `n_clicks`, so I thought what I had was simpler.